routes/booking.js

```diff
@@ -0,0 +1,341 @@
+require("express-async-errors");
+const express = require("express");
+const { Booking } = require("../models/booking");
+const { Room } = require("../models/room");
+const isClient = require("../middleware/isClient");
+const isManager = require("../middleware/isManager");
+const PdfPrinter = require("pdfmake");
+const router = express.Router();
+
+router.post("/requestToBook/", isClient, async (req, res) => {
+  const { arrivalDate, departureDate, roomId } = req.query;
+  const { user } = req;
+
+  if (!arrivalDate || !departureDate || !roomId || !user._id)
+    return res.status(400).send(`Invalid Request`);
+
+  const room = await Room.findOne({ _id: roomId });
+  if (!room) res.status(400).send(`The room does not exists !`);
+
+  const overlappingDatesFilter = {
+    arrivalDate: { $lt: departureDate },
+    departureDate: { $gt: arrivalDate },
+    roomId,
+    isApproved: true,
+  };
+  const existingBooking = await Booking.findOne(overlappingDatesFilter);
+
+  if (existingBooking)
+    return res.status(400).send(`The room is not available on those dates !`);
+
+  const bookingRequest = {
+    userId: user._id,
+    roomId,
+    arrivalDate,
+    departureDate,
+  };
+  const booking = new Booking(bookingRequest);
+
+  const newBooking = await booking.save();
+  res.status(201).send(newBooking);
+});
+
+router.delete("/requestToBook/:bookingId", isClient, async (req, res) => {
+  const { bookingId } = req.params;
+  if (!bookingId) return res.status(400).send(`Invalid Request`);
+
+  const bookingBody = {
+    _id: bookingId,
+  };
+  const existingBooking = await Booking.findOne(bookingBody);
+
+  if (!existingBooking)
+    return res
+      .status(400)
+      .send(`The booking with the id: ${bookingId} does not exists  !`);
+
+  const deletedBooking = await Booking.deleteOne(bookingBody);
+  res.status(200).send(deletedBooking);
+});
+
+router.get("/requestToBook/:bookingId", isClient, async (req, res) => {
+  const { bookingId } = req.params;
+  if (!bookingId) return res.status(400).send(`Invalid Request`);
+
+  const queryBody = {};
+
+  queryBody.bookingBody = { _id: bookingId };
+
+  queryBody.populateUserId = {
+    path: "userId",
+    select: "firstName lastName email",
+  };
+
+  queryBody.populateRoomId = {
+    path: "roomId",
+    select: "type number description",
+  };
+
+  const booking = await Booking.findOne(queryBody.bookingBody)
+    .populate(queryBody.populateUserId)
+    .populate(queryBody.populateRoomId);
+
+  if (!booking)
+    return res
+      .status(400)
+      .send(`The booking with the id: ${bookingId} does not exists  !`);
+  booking;
```

Comment:
Status should be 404 and the bookingId should not be included in the response for security reasons.